### PR TITLE
Add memory bandwidth exporter for AI workload.

### PR DESCRIPTION
**Pull request by Yugar-1**
## Description

This PR adds memory bandwidth exporter for AI workload. Memory bandwidth exporter provides users memory bandwidth metrics of their running containers. The metrics include llc_occupancy, mbm_local_bytes, mbm_total_bytes, cpu utilization and memory usage, and the metrics have been processed. In addition to container-level metrics, it also provides class-level and socket-level metrics. Users can configure the list of metrics to be collected. It serves as an exporter which can be connected to Promethus-like obserbility tools. And it also can be used as a telementry provider. 
Memory bandwidth exporter makes use of state-of-the-art technologies like NRI to build a resource-efficient and well-maintained solution. This solution provides observability to memory bandwidth to OPEA micro-services. It can be deployed separately on end user environments, supporting any cases that memory bandwidth metrics are required.

## Issues

List the issue or RFC link this PR is working on. If there is no such link, please mark it as `n/a`.

n/a

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

List the newly introduced 3rd party dependency if exists.

n/a

## Tests

Describe the tests that you ran to verify your changes.

n/a

### CODE REVIEW

File: kubernetes-addons/memory-bandwidth-exporter/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o memory-bandwidth-exporter cmd/main.go
+
+FROM ubuntu:22.04
+USER root
+WORKDIR /
+# RUN apt-get update && apt-get install -y jq && \
+#     mkdir -p /etc/containers/oci/hooks.d/ 
+
+# COPY --from=builder /workspace/plugin/etc/containers/oci/hooks.d/* /etc/containers/oci/hooks.d/
+# COPY --from=builder /workspace/plugin/usr/local/sbin/* /usr/local/sbin/
```
Comment (daisy-ycguo) [inline kubernetes-addons/memory-bandwidth-exporter/Dockerfile:18]: Remove the commented lines.

File: kubernetes-addons/memory-bandwidth-exporter/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+$ wget https://github.com/containerd/containerd/releases/download/v1.7.0/containerd-1.7.0-linux-amd64.tar.gz
+
+# stop running containerd
+$ sudo systemctl stop containerd
+
+# replace old containerd
+$ sudo tar Cxzvf /usr/local containerd-1.7.0-linux-amd64.tar.gz
+
+# enable NRI in containerd
+add an item in /etc/containerd/config.toml
```
Comment (daisy-ycguo) [inline kubernetes-addons/memory-bandwidth-exporter/README.md:22]: I guess there should be a comment character "#" before this line.

File: kubernetes-addons/memory-bandwidth-exporter/config/prometheus/prometheus.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+global:
+  scrape_interval: 15s
```
Comment (daisy-ycguo) [inline kubernetes-addons/memory-bandwidth-exporter/config/prometheus/prometheus.yml:5]: I don't see where this prometheus.yaml file is used in this feature.

**Review by daisy-ycguo — APPROVED** [review]
unit tests will be added later.

**Review by yongfengdu — APPROVED** [review]

File: kubernetes-addons/memory-bandwidth-exporter/cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		)
+	}
+
+	return handler, nil
+}
+
+func main() {
+	promlogConfig := &promlog.Config{}
+	flag.AddFlags(kingpin.CommandLine, promlogConfig)
+	kingpin.Version(version.Print("memory bandwidth exporter"))
```
Comment (mkbhanda) [inline kubernetes-addons/memory-bandwidth-exporter/cmd/main.go:181]: is there a MBE version number? this print statement mentions no number. Assign it something.

File: kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
+package collector
```
Comment (mkbhanda) [inline kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go:1]: Perhaps want to fix this file name .. guessing you intended it to be "class_collector.go

File: kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+			c.nodeName,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			classLLCacheDesc,
+			prometheus.GaugeValue,
+			stats.processedStats.SumCache.LLCOccupancy,
+			cid,
+			c.nodeName,
+		)
+		for sid, s := range stats.processedStats.MemoryBandwidth {
```
Comment (mkbhanda) [inline kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go:158]: a few comments nice .. what is sid? guessing cid is container-id

File: kubernetes-addons/memory-bandwidth-exporter/collector/collector.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+// NewCollector creates a new Collector.
+func NewCollector(logger log.Logger, interval time.Duration, filters ...string) (*MBCollector, error) {
+	f := make(map[string]bool)
+	for _, filter := range filters {
+		enabled, exist := collectorState[filter]
+		if !exist {
+			return nil, fmt.Errorf("missing collector: %s", filter)
+		}
+		if !*enabled {
+			return nil, fmt.Errorf("disabled collector: %s", filter)
```
Comment (mkbhanda) [inline kubernetes-addons/memory-bandwidth-exporter/collector/collector.go:102]: is this an error? A collector with no filters?

**Review by mkbhanda — COMMENTED** [review]
Where are we checking that this collector is not launched on a node that is not an Intel processor with RDT etc capability. is the memory bandwidth information available on other platforms too. Its OK to have a collector specific to Intel platforms, or a base class with generic stuff and then Intel specific.

Its nice to have multiple distinct ccollectors .. but any chance there is overlap in the metrics they collect and inefficiencies from that?

Nice to have a doc file describing the value of this collector at a high level, the type of insights it can provide -- noisy neighbors, need for more memory/cache for greater performance etc. Potentially a few links to external resources that people can learn more from. Would this be useful for optimization teams, for scheduling decisions ...

File: kubernetes-addons/memory-bandwidth-exporter/collector/collector.go (hunk 7 above)
Comment (mkbhanda) [reply kubernetes-addons/memory-bandwidth-exporter/collector/collector.go:102]: Do you want to add a comment about performance overhead, like O(N) in the number of containers, some K for the filters (llc, memory local ..), some m for periodicity and amount of space needed on disk to hold metrics before it is pumped out.

File: kubernetes-addons/memory-bandwidth-exporter/README.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+# get memory bandwidth metrics
+$ curl http://localhost:9100/metrics
+```
+
+#### How to build the docker image and setup?
+
+```sh
+$ make docker.build
+$ sudo docker run \
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/README.md:73]: Yes, it should be a daemon set that runs on every node, and the node name is a label of the metrics.

File: kubernetes-addons/memory-bandwidth-exporter/README.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+#### How to delete binary?
+
+```sh
+$ make clean
+```
+
+## More flags about memory bandwidth exporter
+
+There are some flags to help users better use memory bandwidth exporter:
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/README.md:124]: Yes, we need to stop and restart if we want to change any of the flag values. Thank you for your suggestion, we will implement it as a feature in the future.

File: kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go (hunk 6 above)
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go:158]: Sorry for the confusion, sid is socket-id and cid is container-id.

File: kubernetes-addons/memory-bandwidth-exporter/cmd/main.go (hunk 4 above)
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/cmd/main.go:182]: Yes, the memory bandwidth exporter has been assigned a version number.

File: kubernetes-addons/memory-bandwidth-exporter/README.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+```
+
+#### How to deploy on the K8s cluster?
+
+Build and push your image to the location specified by `MBE_IMG`, and apply manifest:
+
+```sh
+make docker.build docker.push MBE_IMG=<some-registry>/opea/memory-bandwidth-exporter:<tag>
+make change_img MBE_IMG=<some-registry>/opea/memory-bandwidth-exporter:<tag>
+kubectl create ns system
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/README.md]: Thanks for your kind reminder, I have added the corresponding instructions.

File: kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go (hunk 5 above)
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/collector/class_clollector.go:1]: Sorry for my carelessness, I have fixed class_collector file name.

**Review by mkbhanda — APPROVED** [review]
LGTM

File: kubernetes-addons/memory-bandwidth-exporter/README.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+--collector.class.metrics="none"              Enable class collector metrics.
+--collector.node.metrics="none"               Enable node collector metrics.
+--web.telemetry-path="/metrics"               Path under which to expose metrics.
+--[no-]web.disable-exporter-metrics           Exclude metrics about the exporter itself (promhttp_*, process_*, go_*).
+--web.max-requests=40                         Maximum number of parallel scrape requests. Use 0 to disable.
+--runtime.gomaxprocs=1                        The target number of CPUs Go will run on (GOMAXPROCS) ($GOMAXPROCS)
+--[no-]web.systemd-socket                     Use systemd socket activation listeners instead of port listeners (Linux only).
+--web.listen-address=:9100 ...                Addresses on which to expose metrics and web interface. Repeatable for multiple addresses.
+--web.config.file=""                          Path to configuration file that can enable TLS or authentication. See: https://github.com/prometheus/exporter-toolkit/blob/master/docs/web-configuration.md
+--collector.interval=3s                       memory bandwidth exporter collect metrics interval
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/README.md:144]: Thanks for your suggestion, I will add this feature in issue#407.

File: kubernetes-addons/memory-bandwidth-exporter/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Yugar-1) [reply kubernetes-addons/memory-bandwidth-exporter/README.md:127]: We will add the feature of "dynamically updating flag parameters" in the issue#407.

**Comment by Yugar-1** [comment]
> Where are we checking that this collector is not launched on a node that is not an Intel processor with RDT etc capability. is the memory bandwidth information available on other platforms too. Its OK to have a collector specific to Intel platforms, or a base class with generic stuff and then Intel specific.
> 
> Its nice to have multiple distinct ccollectors .. but any chance there is overlap in the metrics they collect and inefficiencies from that?
> 
> Nice to have a doc file describing the value of this collector at a high level, the type of insights it can provide -- noisy neighbors, need for more memory/cache for greater performance etc. Potentially a few links to external resources that people can learn more from. Would this be useful for optimization teams, for scheduling decisions ...

We will add checks in the collector code to ensure the host node is an Intel platform that supports RDT etc., tracked by issue#406.

The metrics collected by these collectors do not overlap and are obtained and processed from different system files, so they will not cause a performance degradation of the memory bandwidth exporter.

Thank you for your suggestion, it's great! We will add a doc file later to describe the value of memory bandwidth exporter at a high level.

**Pull request merged**